x/staking/querier/querier.go

```diff
@@ -128,14 +129,47 @@ func NewQueryRedelegationParams(delegatorAddr sdk.AccAddress, srcValidatorAddr s
 	}
 }
 
-func queryValidators(ctx sdk.Context, cdc *codec.Codec, k keep.Keeper) (res []byte, err sdk.Error) {
+func queryValidators(ctx sdk.Context, cdc *codec.Codec, req abci.RequestQuery, k keep.Keeper) ([]byte, sdk.Error) {
+	var params QueryValidatorsParams
+
+	err := cdc.UnmarshalJSON(req.Data, &params)
+	if err != nil {
+		return nil, sdk.ErrInternal(fmt.Sprintf("failed to parse params: %s", err))
+	}
+
 	stakingParams := k.GetParams(ctx)
-	validators := k.GetValidators(ctx, stakingParams.MaxValidators)
+	if params.Limit == 0 {
+		params.Limit = int(stakingParams.MaxValidators)
+	}
+
+	validators := k.GetAllValidators(ctx)
+	filteredVals := make([]types.Validator, 0, len(validators))
+
+	for _, val := range validators {
+		if strings.ToLower(sdk.BondStatusToString(val.GetStatus())) == strings.ToLower(params.Status) {
+			filteredVals = append(filteredVals, val)
+		}
+	}
+
+	// get pagination bounds
+	start := (params.Page - 1) * params.Limit
+	end := params.Limit + start
+	if end >= len(filteredVals) {
+		end = len(filteredVals)
+	}
 
-	res, errRes := codec.MarshalJSONIndent(cdc, validators)
+	if start >= len(filteredVals) {
+		// page is out of bounds
+		filteredVals = []types.Validator{}
```

Comment:
should we not directly error here? (- if we did then we unindent everything in the following else statement)

Reply:
Hmmm, I don't think. Asking for a page beyond the total should just give you no results. This is generally how I've seen pagination work (eg. Tendermint also does this).

Reply:
👍